src/util/encoding_rlp.hpp

```diff
@@ -13,129 +13,215 @@ namespace taraxa::util::encoding_rlp {
+void rlp(RLPEncoderRef encoding, std::pair<T1, T2> const& target) {
+  rlp_tuple(encoding, target.first, target.second);
 }
 
-template <typename Map>
-auto enc_rlp(RLPStream& rlp, Map const& target)
-    -> decltype(target.size(), target.begin()->first, target.end()->second, void()) {
-  rlp.appendList(target.size());
-  for (auto const& [k, v] : target) {
-    enc_rlp_tuple(rlp, k, v);
+template <typename Sequence>
+auto rlp(RLPEncoderRef encoding, Sequence const& target)
+    -> decltype(target.size(), target.begin(), target.end(), void()) {
+  encoding.appendList(target.size());
+  for (auto const& v : target) {
+    rlp(encoding, v);
   }
 }
 
 template <typename Param, typename... Params>
-void enc_rlp(RLPStream& rlp, Param const& target, Params const&... rest) {
-  enc_rlp(rlp, target);
-  enc_rlp(rlp, rest...);
+void __enc_rlp_tuple_body__(RLPEncoderRef encoding, Param const& target, Params const&... rest) {
+  rlp(encoding, target);
+  if constexpr (sizeof...(rest) != 0) {
+    __enc_rlp_tuple_body__(encoding, rest...);
+  }
 }
 
 template <typename... Params>
-void enc_rlp_tuple(RLPStream& rlp, Params const&... args) {
-  rlp.appendList(sizeof...(args));
-  enc_rlp(rlp, args...);
+void rlp_tuple(RLPEncoderRef encoding, Params const&... args) {
+  constexpr auto num_elements = sizeof...(args);
+  static_assert(0 < num_elements);
+  encoding.appendList(num_elements);
+  __enc_rlp_tuple_body__(encoding, args...);
 }
 
 template <typename... Params>
-void dec_rlp_tuple(RLP const& rlp, Params&... args);
+void rlp_tuple(RLPDecoderRef encoding, Params&... args);
 
-template <typename Param>
-auto dec_rlp(RLP const& rlp, Param& target) -> decltype(Param(rlp), void()) {
-  target = Param(rlp);
+template <typename T>
+auto rlp(RLPDecoderRef encoding, T& target) -> decltype(encoding.target.toInt<T>(), void()) {
+  target = encoding.target.toInt<T>(encoding.strictness);
 }
 
+template <unsigned N>
+void rlp(RLPDecoderRef encoding, FixedHash<N>& target) {
+  target = encoding.target.toHash<FixedHash<N>>(encoding.strictness);
+}
+
+inline auto rlp(RLPDecoderRef encoding, string& target) { target = encoding.target.toString(encoding.strictness); }
+
+inline auto rlp(RLPDecoderRef encoding, bytes& target) { target = encoding.target.toBytes(encoding.strictness); }
+
 template <typename Param>
-void dec_rlp(RLP const& rlp, optional<Param>& target) {
-  if (rlp.isNull() || rlp.isEmpty()) {
+void rlp(RLPDecoderRef encoding, optional<Param>& target) {
+  if (encoding.target.isNull() || encoding.target.isEmpty()) {
     target = nullopt;
   } else {
-    dec_rlp(rlp, target.emplace());
+    rlp(encoding, target.emplace());
   }
 }
 
 template <typename Sequence>
-void dec_rlp_sequence(RLP const& rlp, Sequence& target) {
-  for (auto const i : rlp) {
-    dec_rlp(i, target.emplace_back());
+auto rlp(RLPDecoderRef encoding, Sequence& target) -> decltype(target.emplace_back(), void()) {
+  for (auto const& i : encoding.target) {
```

Comment:
remove & from here also up in few more loop like this. This will trigger warning under clang  (https://github.com/Taraxa-project/taraxa-node/pull/794)

Reply:
ok

Reply:
done